linkerd/app/core/src/trace.rs

```diff
@@ -7,20 +7,32 @@ use tracing_subscriber::{
     reload, EnvFilter, FmtSubscriber,
 };
 
-const ENV_LOG: &str = "LINKERD2_PROXY_LOG";
+const ENV_LOG_LEVEL: &str = "LINKERD2_PROXY_LOG";
+const ENV_LOG_FORMAT: &str = "LINKERD2_PROXY_LOG_FORMAT";
 
-type Subscriber = Formatter<format::DefaultFields, format::Format<format::Full, Uptime>>;
+const DEFAULT_LOG_LEVEL: &str = "warn,linkerd2_proxy=info";
+const DEFAULT_LOG_FORMAT: &str = "PLAIN";
+
+type JsonFormatter = Formatter<format::JsonFields, format::Format<format::Json, Uptime>>;
+type PlainFormatter = Formatter<format::DefaultFields, format::Format<format::Full, Uptime>>;
 
 #[derive(Clone)]
-pub struct LevelHandle {
-    inner: reload::Handle<EnvFilter, Subscriber>,
+pub enum LevelHandle {
+    Json {
+        inner: reload::Handle<EnvFilter, JsonFormatter>,
+    },
+    Plain {
+        inner: reload::Handle<EnvFilter, PlainFormatter>,
+    },
```

Comment:
@hawkw Do you have any thoughts on this approach over, say, making the enum just the formatter?

i.e.

```rust
enum Formatter {
  Json(JsonFormatter),
  Default(DefaultFormatter),
}
```

and then using `reload::Handle<EnvFilter, Formatter>` throughout?

Is there a reason that we have to vary the whole handle? I don't think we should support changing formats at runtime, for instance...

Reply:
@olix0r if the formatter was an enum, we would need an implementation of the [`FormatEvent`](https://docs.rs/tracing-subscriber/0.2.5/tracing_subscriber/fmt/trait.FormatEvent.html) trait that matches the enum and calls the `FormatEvent` impl for each variant. This enum matching would happen every time we log an event (a relatively hot path). By making the reload handle an enum instead, we only need to match on variants when reloading the level filter (which happens very infrequently). 

> Is there a reason that we have to vary the whole handle? I don't think we should support changing formats at runtime, for instance...

The current implementation doesn't allow that --- the reload layer just needs to be parameterized over the inner subscriber type for type system reasons. If we wanted to change the formatter at runtime, it would need a separate reload layer as well.

Reply:
makes sense, thanks.

Reply:
In an ideal world, the handle wouldn't need to be parameterized over the inner subscriber, just the type of the layer that's actually being reloaded. IIRC, there was some reason this didn't work when this code was originally written. I'd like to try and fix it in `tracing-subscriber` 0.3...